lib/fritzbox/smarthome/actor.rb

```diff
@@ -4,36 +4,64 @@ class Actor < Resource
       include ActiveModel::Model
 
       attr_accessor \
-        :id,
-        :type,
-        :ain,
-        :present,
-        :name,
-        :manufacturer,
-        :group_members
+      :id,
+      :type,
+      :ain,
+      :present,
+      :name,
+      :manufacturer,
+      :group_members
+
+      ResourceNotFound = Class.new(RuntimeError)
 
       class << self
         def all(types: ['group', 'device'])
-          response = get(command: 'getdevicelistinfos')
-          xml = nori.parse(response.body)
+          xml = parse(get(command: 'getdevicelistinfos'))
+
           Array.wrap(types.map { |type| xml.dig('devicelist', type) }.flatten).compact.map do |data|
             klass = Actor.descendants.find { |k| k.match?(data) } || Actor
             self.in?([klass, Actor]) ? klass.new_from_api(data) : nil
           end.compact
         end
 
+        def find_by!(id: nil, ain: nil)
+          raise ArgumentError, 'Find via :id is not implmentet yet' if id.present?
+          raise ArgumentError, 'Missing argument :ain' if ain.nil?
+
+          data = parse(get(command: 'getdeviceinfos', ain: ain)).fetch('device')
+          klass = Actor.descendants.find { |k| k.match?(data) } || Actor
+
+          instance = klass.new(ain: ain)
+          instance.assign_from_api(data)
+          instance
+        end
+
         def new_from_api(data)
-          new(
-            id:            data.dig('@id').to_s,
-            type:          data.dig('groupinfo').present? ? :group : :device,
-            ain:           data.dig('@identifier').to_s,
-            present:       data.dig('present') == '1',
-            name:          (data.dig('name') || data.dig('@productname')).to_s,
-            manufacturer:  (data.dig('manufacturer') || data.dig('@manufacturer')).to_s,
-            group_members: data.dig('groupinfo', 'members').to_s.split(',').presence
-          )
+          instance = new
+          instance.assign_from_api(data)
+          instance
         end
       end
+
+      def assign_from_api(data)
+        assign_attributes(
+          id:            data.dig('@id').to_s,
+          type:          data.dig('groupinfo').present? ? :group : :device,
+          ain:           data.dig('@identifier').to_s,
+          present:       data.dig('present') == '1',
+          name:          (data.dig('name') || data.dig('@productname')).to_s,
+          manufacturer:  (data.dig('manufacturer') || data.dig('@manufacturer')).to_s,
+          group_members: data.dig('groupinfo', 'members').to_s.split(',').presence
+        )
+      end
+
+      def reload
+        xml = parse(get(command: 'getdeviceinfos', ain: ain))
+        assign_from_api(xml.fetch('device'))
+        self
+      rescue KeyError
+        raise ResourceNotFound
```

Comment:
:question:  **Optional**

Would also return the `ain` of the resource:

```suggestion
        raise ResourceNotFound.new("Unable to reload self with ain #{ain.inspect}")
```

Reply:
Makes sense.